components/global/core/io.cellery.observability.telemetry.deduplicator/src/main/java/io/cellery/observability/telemetry/deduplicator/DeduplicationWindowExtension.java

```diff
@@ -0,0 +1,403 @@
+                                lastTimestamp = clonedEvent.getTimestamp();
+                            }
+                        }
+                    }
+                    expiredEventChunk.reset();
+                    if (streamEvent.getType() != StreamEvent.Type.CURRENT) {
+                        streamEventChunk.remove();
+                    }
+                }
+            }
+            nextProcessor.process(nextProcessorEventChunk);
+    }
+
+    private boolean checkForDuplications(StreamEvent oldEvent, StreamEvent currentEvent) {
+        String oldEventTraceId = (String) traceIdExecutor.execute(oldEvent);
+        String oldEventSpanId = (String) spanIdExecutor.execute(oldEvent);
+        String oldEventParentSpanId = (String) parentSpanIdExecutor.execute(oldEvent);
+        String oldEventSourceNamespace = (String) sourceNamespaceExecutor.execute(oldEvent);
+        String oldEventSourceInstance = (String) sourceInstanceExecutor.execute(oldEvent);
+        String oldEventSourceComponent = (String) sourceComponentExecutor.execute(oldEvent);
+        String oldEventDestinationNamespace = (String) destinationNamespaceExecutor.execute(oldEvent);
+        String oldEventDestinationInstance = (String) destinationInstanceExecutor.execute(oldEvent);
+        String oldEventDestinationComponent = (String) destinationComponentExecutor.execute(oldEvent);
+
+        String currentEventTraceId = (String) traceIdExecutor.execute(currentEvent);
+        String currentEventSpanId = (String) spanIdExecutor.execute(currentEvent);
+        String currentEventParentSpanId = (String) parentSpanIdExecutor.execute(currentEvent);
+        String currentEventSourceNamespace = (String) sourceNamespaceExecutor.execute(currentEvent);
+        String currentEventSourceInstance = (String) sourceInstanceExecutor.execute(currentEvent);
+        String currentEventSourceComponent = (String) sourceComponentExecutor.execute(currentEvent);
+        String currentEventDestinationNamespace = (String) destinationNamespaceExecutor.execute(currentEvent);
+        String currentEventDestinationInstance = (String) destinationInstanceExecutor.execute(currentEvent);
+        String currentEventDestinationComponent = (String) destinationComponentExecutor.execute(currentEvent);
+
+        return currentEventSourceNamespace.equals(oldEventSourceNamespace)
+                && currentEventSourceInstance.equals(oldEventSourceInstance)
+                && currentEventSourceComponent.equals(oldEventSourceComponent)
+                && currentEventDestinationNamespace.equals(oldEventDestinationNamespace)
+                && currentEventDestinationInstance.equals(oldEventDestinationInstance)
+                && currentEventDestinationComponent.equals(oldEventDestinationComponent)
+                && oldEventTraceId.equals(currentEventTraceId)
+                && (oldEventSpanId.equals(currentEventParentSpanId) || oldEventParentSpanId.equals(currentEventSpanId));
```

Comment:
IINM `oldEventParentSpanId` can be null. Shall we use Objects.equls() method for all the equals checks in this condition to be on the safe side ?